### ISSUE TITLE
Handle Disabled LocalStorage in Browser and substitute an in-memory provider

### ISSUE THREAD

**Issue by csharpfritz** [Blazored/LocalStorage#188]
**Is your feature request related to a problem? Please describe.**
If a user has disabled cookies in their browser, all local storage access throws an error.

**Describe the solution you'd like**
Can we provide an alternate in-memory storage provider instead of using the browser localStorage so that code will still work, but not persist between sessions.

**Comment by chrissainty** [comment]
Hey @csharpfritz! 

Thanks for raising this issue. I didn't realise local storage was disabled along with cookies--you learn something new everyday! 

Yes, I think swapping to an in-memory option would be possible. It's what we do for the test extensions already. 

I think this should be a setting developers turn on, rather than being a default. Some new API like, 'EnableInMemoryStorageFallback'. We can then use the result of the 'navigator.cookieEnabled' JavaScript API to enable the in-memory provider.

**Comment by csharpfritz** [comment]
I like the idea of a configuration option, but I think if localStorage isn’t available, developers should have an appropriate exception thrown



Jeff

**Comment by chrissainty** [comment]
Interesting, what use would the exception have for developers? The issue will only surface when the application is running on end users devices. Throwing an exception at that point would potentially kill the application, depending on how the developer has designed their code.

Wouldn't this be a feature a developer would enable ahead of time in order to keep things running if the end user has disabled cookies?

**Comment by csharpfritz** [comment]
I'm referring to an exception if AppSettings is inaccessible and the In-Memory fallback is not enabled.

"AppSettingsUnavailableException" or similar

**Comment by joepvtl** [comment]
Hi,

This would be a great addition! Am I correct to assume that data in the InMemory fallback only exists untill the user reloads the app?

**Comment by csharpfritz** [comment]
Yes, in a WebAssembly scenario all context is lost when the app is reloaded. 

For users who are privacy concerned, this should be a fair consideration in exchange for not writing any content to the local disk.

**Comment by chrissainty** [comment]
Sorry this has taken so long to make any progress. I've added in a new `BrowserStorageDisabledException` that will through if the browsers storage is unavailable. For now, I've decided not to implement a fall back. This is due to the fact it will behave differently to how local storage, i.e. not persistent. I'll create a new issue for the fallback and see what people feeling are on it.
